### PR TITLE
Add -verror-format-level and extended format strings + errorEx

### PR DESCRIPTION
**Pull request by BradleyChatha**
## What does this do?

Tl;dr introduces a new command argument `-verror-format-level=[0|1|2|etc..]`to allow configuration of how certain error messages may be displayed, using a new 'extended' format string.

Given this code:

```d
struct S(alias Func){ }

int func1(int a){ return a*2; }
int func2(int a){ return a*2; }

void main()
{
   auto a = S!func1();
   auto b = S!func2();

   a = b;

   auto c = S!((int a) => a*2)();
   auto d = S!((int a) => a*2)();

   c = d;
}
```

The code generating the error message:
```d
auto ts = toAutoQualChars(e.type, t);
errorEx(e.loc, "cannot implicitly convert expression `%s` of type $[indent:2]{`%s`} $[indent:1]{to} $[indent:2]{`%s`}",
    e.toChars(), ts[0], ts[1]);
```

With the default message level (0) we get this output:

```
test.d(11): Error: cannot implicitly convert expression `b` of type `S!(func2)` to `S!(func1)`
test.d(16): Error: cannot implicitly convert expression `d` of type `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)` to `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)`
```

With message level 1, the error is split onto multiple lines:

```
test.d(11): Error: cannot implicitly convert expression `b` of type
test.d(11):             `S!(func2)`
test.d(11):        to
test.d(11):             `S!(func1)`
test.d(16): Error: cannot implicitly convert expression `d` of type
test.d(16):             `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)`
test.d(16):        to
test.d(16):             `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)`
```

This PR only implements the initial ground work for formatting levels, as well as only implementing the `$[indent]` formatter.

Further formatters such as the "theoretical level 3" one described below will be left to separate PRs.

## Code additions

This PR adds the following functions into `errors.d`:

```d
extern (C++) void errorEx(const ref Loc loc, const(char)* format, ...)
extern (C++) void verrorFormatPrint(const ref Loc loc, const(char)* format, va_list ap, MessageFunc print, MessageFunc printSupplemental)

// These two can mostly be ignored, as they're not likely to be used on their own
private const(char)* verrorFormatPrefixString(const(char)* format, va_list ap)
private bool verrorFormatNextFragment(const(char)* prefixedFormat, out FragmentInfo info, ref const(char)* nextPrefixedFormat)
```

The bottom two functions are mostly subroutines for `verrorFormatPrint`, though they can be used separately if needed.

`verrorFormatPrint` supports an 'extended' version of a printf-style format string, which allows for 'formatting fragments' to be used to format output strings differently, depending on the error format level set by the user.

`verrorFormatPrint` does not actually perform any printing by itself, but instead outsources that job to its `print` and `printSupplemental` parameters, to allow for easy integration.

The 'extended' format string is a normal printf compatible format string, except it allows for statements such as: `$[FORMATTER:PARAMS]{INNER_BODY}` (e.g. `${indent:1}{At message level 1, I'm on a new line and indented by 1 space-tab!}`) can be used.

`indent` is the only formatter this PR will implement, and is documented in the code.

To support certain kinds of formatters, all formatters are ran *after* expansion of the printf part of the string. e.g. ` "$[indent:0]{%s}" -> $[indent:0]{Hello world!}" -> Formatters are then ran.`

Finally, `errorEx` is simply the 'extended format' version of the normal `error` function.

## Other additions

This PR applies changes to the `cannot implicitly convert expression '...' of type '...' to '...'` error message, such that the two different types are put onto separate lines if the format level is 1 or greater.

This PR adds a new compiler flag: `-verror-format-level=[0|1|2|3]` which is used to set the format level. Level 0 is default formatting, and as the levels increase, the more "involved"/"extreme" the formatting becomes.

There's 4 different levels defined, but only 2 are used right now (0 and 1) just for a slight bit of future proofing. Levels can be added and changed and whatnot of course.

## Further considerations

In the current form for this PR, format levels are only used by formatters, which simply run over pre-made strings, and cannot actually provide any further information within them (e.g. going from `int function()` to `mod1.func1.__lambda1 is int function()`), so perhaps in the future other parts of the code base to do with generating user-displayed strings will make use of the format level.

The current implementation of `verrorFormatPrint` does not support formatters being embedded within other formatters, which will likely be an annoying limitation for pieces of text that may want to allow different formatters for different format levels to apply to it. This can easily be changed in the future though. e.g. we could support embedded formatters, or add additional syntax such as `$[formatter1:params1][formatter2:param2]{inner text}`, or whatever, really.

(I don't mind implementing the above, but I didn't want to complicate things too much initially.)

Is this the right solution? I don't know, especially since the initial PR was quite a bit different. Can it be made better? Definitely. Is it at the very least better than nothing? Certainlyish.

## Motivation & References

There's recently been two threads on the learning group [1] [2] where there's been issues regarding the extremely type names for templated types.

There's been many complaints in general about how tricky it can be to parse error messages relating to templates.

So what justification is there to not provide opt-in ways of making these messages easier for the user to parse through them?

This is mostly the start of more sophisticated formatting, as only a single formatter (`$[indent]` at level 1) is implemented.

[1] https://forum.dlang.org/thread/vypnzuhggxyqmqrurmrh@forum.dlang.org
[2] https://forum.dlang.org/thread/mbgbirwallrrectcvuxg@forum.dlang.org

## Problems

It'd probably take me a decent amount of time and question asking in write code that's actually usable in anyway, so I don't know if I can actually properly implement this myself. 

### Why a command line option?

Because I fear making any one of these changes the default might not be acceptable, so having the various levels become opt-in should make it more reasonable to add in.

### Why at all?

Let's take the error message from [1]:

```
das2/range.d(359,31): Error: constructor das2.range.PriorityRange!(DasRange!(Tuple!(int, int)[], int function(Tuple!(int, int)) pure nothrow @nogc @safe, int function(Tuple!(int, int)) pure nothrow @nogc @safe, Tuple!(int, int), int), int function() pure nothrow @nogc @safe).PriorityRange.this(DasRange!(Tuple!(int, int)[], int function(Tuple!(int, int)) pure nothrow @nogc @safe, int function(Tuple!(int, int)) pure nothrow @nogc @safe, Tuple!(int, int), int) range, int function(Tuple!(int, int)) priority) is not callable using argument types (DasRange!(Tuple!(int, int)[], int function(Tuple!(int, int)) pure nothrow @nogc @safe, int function(Tuple!(int, int)) pure nothrow @nogc @safe, Tuple!(int, int), int), int function() pure nothrow @nogc @safe)
```

This user was using D in a perfectly fine and normal way, yet is punished with having to manually parse/format these horrific blobs of text:

```
PriorityRange.this(
	DasRange!(
		Tuple!(int, int)[],
		int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		Tuple!(int, int),
		int
	) range,
	int function(Tuple!(int, int)) priority
) is not callable using argument types (
	DasRange!(
		Tuple!(int, int)[],
		int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		Tuple!(int, int),
		int
	),
	int function(Tuple!(int, int))
)
```

We also now come to the second issue: The above message on initial glance seems like a bug - the two types are identical, right?

Wrong, if a theoretical level 3 formatter was turned on, we'd might instead get something like:

```
PriorityRange.this(
	DasRange!(
		...,
		mymod.myfunc.__lambda1 is int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		mymod.myfunc.__lambda2 is function(Tuple!(int, int)) pure nothrow @nogc @safe,
		...,
		...
	),
	...
) is not callable using argument types (
	DasRange!(
		...,
		mymod.myfunc.__lambda3 is int function(Tuple!(int, int)) pure nothrow @nogc @safe,
		mymod.myfunc.__lambda4 is function(Tuple!(int, int)) pure nothrow @nogc @safe,
		...,
		...
	),
	...
)
```

Where we can see where the actual discrepancy is between the types.

### CODE REVIEW

**Comment by adamdruppe** [comment]
I like this a lot.

**Comment by TungstenHeart** [comment]
I think that level 1 is worth a try but the current design has problems.

The way it is done now would create a lot of new code in many files (like the `final void errorNoImplicitConvert(Type t)` you have written). 

That should rather be done in the error module, using new a new function that would be able to format using the arguments, to level 0 or level 1.

So let's say

1. define the new error functions, that are able to output both level, use them  to replace cast.d errors
2. during following monthes adapt more and more errors

The new error function could use a system of fragments (by line ?):

so instead of 

```d
e.error("cannot implicitly convert expression `%s` of type `%s` to `%s`", e.toChars(), ts[0], ts[1]);
```

the new function, generic enough could be used as the old one, although with more arguments

```d
e.errorEx( indentErrLine!0("cannot implicitly convert expression `%s` of type", e.toChars())
    indentErrLine!1("`%s`", ts[0]), indentErrLine!0(, "to type"), indentErrLine!1("`%s`", ts[1],),
);
```

and errorEx would be able to build the error in a single line or on several.

```d
errorEx(T...)(T t) {}
```
`indentErrLine` would be a function that returns an indented fragment (as struct with for member a specifier, and array of `RootObject` for the args ?). More fragment types would be added in the future.

It would be the job of `errorEx` to handle them properly, typically in a `static foreach` loop. This way no new error routines would have to be added for each context.

Having a generic solution is also better for the coverage. This way the tests will not have to duplicated to cover each new context specific error function.

**Comment by BradleyChatha** [comment]
Yea, I've been thinking for a bit now on how the actual implementation should look like instead of this bashed together one.

I think I like your idea quite a lot though, so I'll try it.

**Comment by BradleyChatha** [comment]
Doc comments and stuff will come once I know I won't need to delete them immediately ;P

**Comment by cpiker** [comment]
> Given this code:
> 
> ```d
> struct S(alias Func){ }
> 
> int func1(int a){ return a*2; }
> int func2(int a){ return a*2; }
> 
> void main()
> {
>    auto a = S!func1();
>    auto b = S!func2();
> 
>    a = b;
> 
>    auto c = S!((int a) => a*2)();
>    auto d = S!((int a) => a*2)();
> 
>    c = d;
> }
> ```
> 
> With the default message level (0) we get this output:
> 
> ```
> test.d(11): Error: cannot implicitly convert expression `b` of type `S!(func2)` to `S!(func1)`
> test.d(16): Error: cannot implicitly convert expression `d` of type `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)` to `test.S!(function (int a) pure nothrow @nogc @safe => a * 2)`
> ```

For this particular example I think the main issue is that lambdas and regular functions are not reported in a consistent manner.  The function signature is expanded for the lambda but hidden for the regular function and the name is hidden for the lambda but displayed for the regular function.  The problem would be more apparent, even at message level (0), if the output was consistent:
```
test.d(11): Error: cannot implicitly convert expression `b` of type `S!(func2)` to `S!(func1)`
test.d(16): Error: cannot implicitly convert expression `d` of type `S!(__lambda2)` to `S!(__lambda1)`
```

**Comment by TungstenHeart** [comment]
The way it's done with latest commits is better. Nest step is pass the test suite tho.

**Comment by Imperatorn** [comment]
Could we make level 1 the default so that you don't have to specify it?

**Comment by BradleyChatha** [comment]
> 
> 
> Could we make level 1 the default so that you don't have to specify it?

We could, I don't know what the consensus on that would be though.

While this PR only modifies a single error message, there will likely be more in the future, so would defaulting to level 1 eventually be too much of a mess?

In other words: Someone else can make that decision, since I don't have the foresight about whether it's a good idea or not :P

**Comment by BradleyChatha** [comment]
> 
> 
> For this particular example I think the main issue is that lambdas and regular functions are not reported in a consistent manner. The function signature is expanded for the lambda but hidden for the regular function and the name is hidden for the lambda but displayed for the regular function. The problem would be more apparent, even at message level (0), if the output was consistent:
> 
> ```
> test.d(11): Error: cannot implicitly convert expression `b` of type `S!(func2)` to `S!(func1)`
> test.d(16): Error: cannot implicitly convert expression `d` of type `S!(__lambda2)` to `S!(__lambda1)`
> ```

I had a look into that, and after going around in circles for what felt like 2 hours I think I finally found where that might be happening. From a quick judgement it seems pretty annoying to change, but I'll have to mess around with it (and probably put it in a different PR if I manage to get anywhere with it).

**Comment by BradleyChatha** [comment]
> 
> 
> The way it's done with latest commits is better. Next step is pass the test suite tho.

From what I can see, my changes aren't causing the test failures, it seems to be a Phobos thing?

**Review by ibuclaw — CHANGES_REQUESTED** [review]
Whatever you end up implementing, the public facing function should be `extern(C++)` if possible. As dmd's diagnostic handlers do not integrate with gdc/ldc, there should be a way for us to implement the routine on our side.

**Comment by BradleyChatha** [comment]
> 
> 
> Whatever you end up implementing, the public facing function should be `extern(C++)` if possible. As dmd's diagnostic handlers do not integrate with gdc/ldc, there should be a way for us to implement the routine on our side.

Definitely something that slipped my mind >:c

Anyway, I think the new code should be friendly for LDC/GDC? Also see all the `// TEMP-PR-NOTE:` comments.

File: src/dmd/errors.d [hunk 1]
```diff
@@ -35,46 +35,190 @@ enum Classification : Color
...
 {
-    return MsgIndentedLine(text, IndentLevel, FormatLevel);
+    foreach (formatter; _preMessageFormatters)
+        if (!strcmp(name, formatter.formatName))
+            return formatter;
+
+    return null;
 }
 
+private extern (C++) class NewLineAndIndentFormatter : PreMessageFormatter
```
Comment (ibuclaw) [inline src/dmd/errors.d:103]: private stuff doesn't need to be `C++`.

File: src/dmd/errors.d [hunk 2]
```diff
@@ -35,13 +35,190 @@ enum Classification : Color
...
+        uint indentLevel;
+        // TEMP-PR-NOTE: Should I even bother with error checking params?
+        if(params)
+            sscanf(params, "%u", &indentLevel);
+        makeNewLine(locChars, buf);
+        makeIndent(indentLevel, buf);
+    }
+}
+
+pragma(printf) extern (C++) void errorEx(const ref Loc loc, const(char)* format, ...)
```
Comment (ibuclaw) [inline src/dmd/errors.d:121]: So this should also be a prototype in the `errors.h` header too.

**Comment by ibuclaw** [comment]
> > Whatever you end up implementing, the public facing function should be `extern(C++)` if possible. As dmd's diagnostic handlers do not integrate with gdc/ldc, there should be a way for us to implement the routine on our side.
> 
> Definitely something that slipped my mind >:c
> 
> Anyway, I think the new code should be friendly for LDC/GDC? Also see all the `// TEMP-PR-NOTE:` comments.

So the idea is that the entrypoints are `extern(C++`), and the internals are private.  On GDC's side, I actually implement these routines in a way that works with GCC's diagnostic machinery.  See [here](https://github.com/D-Programming-GDC/gcc/blob/4d46018791ee0331b2d867b2420ca566ea4c3a8f/gcc/d/dmd/errors.d#L20-L66) for the functions that are used ad-verbatim, and [here](https://github.com/D-Programming-GDC/gcc/blob/4d46018791ee0331b2d867b2420ca566ea4c3a8f/gcc/d/dmd/errors.d#L191-L210) for what are just D prototypes whose actual implementations are [here](https://github.com/D-Programming-GDC/gcc/blob/4d46018791ee0331b2d867b2420ca566ea4c3a8f/gcc/d/d-diagnostic.cc#L157-L199).

**Comment by ibuclaw** [comment]
To avoid duplicating too much logic into GDC, the best thing would be to have as much of the formatting encapsulated, and the function that deals will the actual printing (calling `verrorPrint`) squirrelled away into its own function.  This function will then be the thing I override with an alternative implementation.

**Comment by BradleyChatha** [comment]
Just so I know I'm understanding things properly (thank you for your patience):

I need to split the formatting code into its own function e.g. `vformatMessage`.

Now my question is, for DMD's error messages, each time a new line is made I need to prefix it with the location, which I assume isn't helpful in GDC's case. But even if that were optional, then it'd probably be a hassle on your end to have to get all the new lines printing properly, right?

So what I'm thinking is, should the format function be constructing a full string to return, or should it move towards something like using a callback anytime it needs to push data?

e.g. `abc $[indent:1] def` -> `push("abc"), pushNewLine(), push("    "), push("def")`

Or generally any other way that'd be easiest for the other compilers to use?

**Comment by RazvanN7** [comment]
cc @WalterBright @atilaneves

**Comment by wilzbach** [comment]
Have we considered how this will work with the existing `-verrors` flags? For example, `-verrors=context`.

Also, I think it would be nice to keep the existing CLI naming scheme and start with `-verror` (e.g. `-verror-format-level`).

See also:
https://dlang.org/dmd-linux.html#switch-verror-style

**Comment by BradleyChatha** [comment]
> 
> 
> Have we considered how this will work with the existing `-verrors` flags? For example, `-verrors=context`.
> 
> Also, I think it would be nice to keep the existing CLI naming scheme and start with `-verror` (e.g. `-verror-format-level`).
> 
> See also:
> https://dlang.org/dmd-linux.html#switch-verror-style

Not at all. I need to bother updating the initial post as the current code essentially makes this all opt-in for each message, instead of just running over the final output willy-nilly.

In other words, anything that doesn't use `errorEx` should be unaffected.

**Comment by BradleyChatha** [comment]
I've updated the initial post slightly to make things more clear and accurate.

Also, if this is somehow not going to be useable, then I honestly don't know what else I could do :(

**Comment by BradleyChatha** [comment]
Also another silly question, how do I run the tests locally?

I'm trying to use `test/run.d` (dmd -run, rdmd, and ci.sh test_dmd) but I get a bunch of errors relating to Phobos regarding `std.format.write.formatValue`, which I assume shouldn't be happening.

Using WSL and DMD 2.096.1.

**Comment by MoonlightSentinel** [comment]
> Also another silly question, how do I run the tests locally?

`rdmd test/run.d` should just work but need a specific setup to execute the the bash tests on windows.
This seems to be a Phobos regression (CC @berni44), could you post the error message / stack trace? 

BTW: You might want to pass `ARGS=` to skip backend permutations

**Comment by BradleyChatha** [comment]
> 
> 
>  could you post the error message / stack trace?

Ran with `rdmd test/run.d &> test.log`

```
Executing: /mnt/d/Coding/D/dmd/dmd/generated/linux/release/64/dmd -conf= -m64 -of/mnt/d/Coding/D/dmd/dmd/test/test_results/dshell_prebuilt.o -c /mnt/d/Coding/D/dmd/dmd/test/tools/dshell_prebuilt/dshell_prebuilt.d -fPIC
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/internal/write.d(2944): Error: cannot implicitly convert expression `val` of type const(FormatSpec!char)* to shared(const(void*))
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1239): Error: template instance `std.format.internal.write.formatValueImpl!(Appender!string, const(FormatSpec!char)*, char)` error instantiating
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(641):        instantiated from here: `formatValue!(Appender!string, const(FormatSpec!char)*, char)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(621):        instantiated from here: `toString!(Appender!string)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(533):        instantiated from here: `FormatSpec!char`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/stdio.d(1814):        ... (1 instantiations, -v to show) ...
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/stdio.d(4444):        instantiated from here: `writefln!(char, string)`
/mnt/d/Coding/D/dmd/dmd/test/tools/dshell_prebuilt/dshell_prebuilt.d(106):        instantiated from here: `writefln!(char, string)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(643): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, const(int), FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(645): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, const(int), FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(647): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, const(char), FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(649): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, const(ubyte), FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(651): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, const(ubyte), FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(653): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, bool, FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(655): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, bool, FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(657): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, bool, FormatSpec!char)`, candidates are:
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/write.d(1231):        `formatValue(Writer, T, Char)(auto ref Writer w, auto ref T val, ref scope const FormatSpec!Char f)`
/mnt/d/Coding/D/dmd/dmd/test/../../phobos/std/format/spec.d(659): Error: template `std.format.write.formatValue` cannot deduce function from argument types `!()(Appender!string, bool, FormatSpec!char)`, candidates are:
failed to build '/mnt/d/Coding/D/dmd/dmd/test/test_results/dshell_prebuilt.o'
```

DMD: `DMD64 D Compiler v2.096.1`

Phobos is on commit: `d7bf7301be88245f4200a26b5ac8d32ec09f35c0`

**Comment by BradleyChatha** [comment]
Five questions (sorry in advance for the annoyance!):

1. What is the standard way to handle errors that only happen during development of the compiler? e.g. In cases where the developer provides a malformed format string, how should it be handled? `assert`, print then exit, something else?
2. Should I write tests for this feature? If so, is there any overview on how to write tests for DMD, especially in regards to checking message output?
3. Is this now suitable for GDC & LDC, or is there something else that's still a no-go outside of the implementation still being unfinished?
4. Should this PR only focus only on adding `$[indent]` as the initial, proof-of-concept fragment, and any further fragments be done separately (assuming this PR is accepted). And should this PR only apply it to a single error message "could not implicitly convert..." as a proof-of-concept sort of thing?
5. While I've documented the actual code, is there anywhere else I should mention this new feature? Docs, changelog, etc?

**Comment by MoonlightSentinel** [comment]
> 1. What is the standard way to handle errors that only happen during development of the compiler? e.g. In cases where the developer provides a malformed format string, how should it be handled? `assert`, print then exit, something else?

Bugs should be caught by appropriate `assert`s.

> 2. Should I write tests for this feature? If so, is there any overview on how to write tests for DMD, especially in regards to checking message output?

Yes. `test/README.md` provides a high level overview while individual test categories are explained in their respective `test/***/README.md`.

A `fail_compilation` test would be appropriate if you are going to apply the new formatting to an existing error. The file should trigger the error and include the expected output in a comment:

```d
/*
TEST_OUTPUT:
---
<Expected output>
---
*/

...
```

Otherwise you could create a `unit` / `dshell` test that uses the new functions directly and redirects the output.

> 4. Should this PR only focus only on adding `$[indent]` as the initial, proof-of-concept fragment, and any further fragments be done separately (assuming this PR is accepted). And should this PR only apply it to a single error message "could not implicitly convert..." as a proof-of-concept sort of thing?

That depends on the complexity of the changes, but I think a single example would be a good start.

**Comment by BradleyChatha** [comment]
Bumping this as I'm just wondering what else is left to do?

I *think* the implementation is finished now (barring bugs of course).

The autotester failed for FreeBSD64_64, but I can't quite see where the actual error is, and whether it's related to my changes or not.

**Comment by thewilsonator** [comment]
try rebasing.

**Comment by BradleyChatha** [comment]
*May have forgotten to squash*

Anyway, that failure is related to vibe (and I'm pretty sure not my code, I hope >:3).

File: src/dmd/errors.d [hunk 3]
```diff
@@ -847,3 +906,215 @@ private void writeHighlights(Console con, ref const OutBuffer buf)
...
+ *                        but it makes it more clear on what you're attempting to do.
+ *
+ * Params:
+ *  loc                 = The location of the error.
+ *  format              = The extended format string to use.
+ *  ap                  = The argument list to use with the `format` string.
+ *  print               = The print function to use for the first line printed.
+ *  printSupplemental   = The print function to use for every line printed after the first line.
+ */
+extern (C++) void verrorFormatPrint(const ref Loc loc, const(char)* format, va_list ap, MessageFunc print, MessageFunc printSupplemental)
```
Comment (ibuclaw) [inline src/dmd/errors.d:942]: I'm struggling to see a reason why this is extern C++.

File: src/dmd/errors.d [hunk 4]
```diff
@@ -35,13 +35,69 @@ enum Classification : Color
...
+ *      loc    = location of error
+ *      format = extended format specification
+ *      ap     = printf-style argument list
+ */
+extern (C++) void verrorEx(const ref Loc loc, const(char)* format, va_list ap)
+{
+    verrorFormatPrint(loc, format, ap, &_verrorExPrint, &_verrorExPrintSupplemental);
+}
+extern (C++) private void _verrorExPrint(const ref Loc loc, const(char)* text) { error(loc, "%s", text); }
+extern (C++) private void _verrorExPrintSupplemental(const ref Loc loc, const(char)* text){ errorSupplemental(loc, "%s", text); }
```
Comment (ibuclaw) [inline src/dmd/errors.d:94]: Also these two functions.  If you are just forwarding to extern(C++) functions now, then I don't see any need to keep these.

Comment (BradleyChatha) [reply src/dmd/errors.d:94]: Ah, good point. I think I was just blindly copying the style of the other error funcs.

File: src/dmd/errors.d (hunk 3 above)
Comment (BradleyChatha) [reply src/dmd/errors.d:942]: Wasn't certain if there'd be a need for it to be extern C++ or not, so I put it on just in case.

Assuming by your answer though that it should stay a private D func.

Comment (ibuclaw) [reply src/dmd/errors.d]: IIRC, your first iteration of this called `verrorPrint` directly, which is the DMD-internal diagnostic routine.

Having a look over of the current iteration, you are now using `error` and `errorSupplemental`, which are the user-facing entrypoints.  If it remains that way, then there's no need to export anything to C++, nor have the function callbacks, as it'll "just work" with whatever back-end diagnostic handler.

File: src/dmd/errors.d [hunk 5]
```diff
@@ -847,3 +904,215 @@ private void writeHighlights(Console con, ref const OutBuffer buf)
...
+ *                        but it makes it more clear on what you're attempting to do.
+ *
+ * Params:
+ *  loc                 = The location of the error.
+ *  format              = The extended format string to use.
+ *  ap                  = The argument list to use with the `format` string.
+ *  print               = The print function to use for the first line printed.
+ *  printSupplemental   = The print function to use for every line printed after the first line.
+ */
+private void verrorFormatPrint(const ref Loc loc, const(char)* format, va_list ap, MessageFunc print, MessageFunc printSupplemental)
```
Comment (ibuclaw) [inline src/dmd/errors.d:940]: Should this be named `vFormatPrint`?  It looks like the `MessageFunc` could be an error, warning, deprecation, tip, etc..

Comment (ibuclaw) [reply src/dmd/errors.d:940]: Ehh, of course it's named after the command-line option, and not the `verror` function itself.

File: src/dmd/errors.d [hunk 6]
```diff
@@ -35,13 +35,67 @@ enum Classification : Color
     tip = Color.brightGreen,          /// for tip messages
 }
 
+alias MessageFunc = extern (C++) void function(const ref Loc loc, const(char)* format, ...);
```
Comment (ibuclaw) [inline src/dmd/errors.d:38]: Private?  Or better yet...

File: src/dmd/errors.d (hunk 5 above)
Comment (ibuclaw) [inline src/dmd/errors.d:940]: ... this could be templated to hide the `MessageFunc` type.
```suggestion
private void verrorFormatPrint(Fun)(const ref Loc loc, const(char)* format, va_list ap, Fun print, Fun printSupplemental)
```

File: src/dmd/errors.h [hunk 7]
```diff
@@ -42,6 +44,10 @@ D_ATTRIBUTE_FORMAT(2, 0) void vmessage(const Loc& loc, const char *format, va_li
 D_ATTRIBUTE_FORMAT(1, 2) void tip(const char *format, ...);
 D_ATTRIBUTE_FORMAT(1, 0) void vtip(const char *format, va_list ap);
 
+// Used when dealing with errorEx-style format strings.
+typedef void (*MessageFunc)(const Loc& loc, const char* text);
+D_ATTRIBUTE_FORMAT(2, 0) void verrorFormatPrint(const Loc& loc, const char* format, va_list ap, MessageFunc print, MessageFunc printSupplemental);
```
Comment (ibuclaw) [inline src/dmd/errors.h:49]: No longer needed.

File: src/dmd/errors.d (hunk 5 above)
Comment (BradleyChatha) [reply src/dmd/errors.d:940]: Don't even remember why I called it that, maybe just as a "eh, it'll do" thing. I can change it if you want though.

**Comment by BradleyChatha** [comment]
I completely forgot to comment.

That CI error is related to vibe-d, not this PR.

And:

Is there anything else left to do for this PR?

**Comment by BradleyChatha** [comment]
I came to close this manually, but didn't realise it's already closed from me deleting my fork, so I apologise for the ping this'll probably cause, but I didn't intend for it to have already been closed...

I'm still semi-interested in looking into something like this at one point; whenever I have enough willpower to attempt such a thing.

Perhaps this isn't really the best place to ask but, would refactoring the idea into a sort of `ErrorBuilder` kind of style be a bit better than this magic string non-sense I for some reason thought was a good idea?

e.g. instead of

```d
errorEx(e.loc, "cannot implicitly convert expression `%s` of type $[indent:1]{`%s`} $[indent:0]{to} $[indent:1]{`%s`}", e.toChars(), ts[0], ts[1]);
```

You'd have (roughly speaking):

```d
ErrorBuilder(e.loc)
  .put("cannot implicitly convert expression `%s` of type ", e.toChars())
  .putVerboseIndent(1, "`%s`", ts[0])
  .putVerboseNewLine("to")
  .putVerboseIndent(1, "`%s`", ts[1])
  .printAsError();
```
